core/engine/src/builtins/intl/number_format/mod.rs

```diff
@@ -1,4 +1,848 @@
+use std::borrow::Cow;
+
+use boa_gc::{Finalize, Trace};
+use boa_macros::utf16;
+use boa_profiler::Profiler;
+use fixed_decimal::{FixedDecimal, FloatPrecision, SignDisplay};
+use icu_decimal::{
+    options::{FixedDecimalFormatterOptions, GroupingStrategy},
+    provider::DecimalSymbolsV1Marker,
+    FixedDecimalFormatter, FormattedFixedDecimal,
+};
+
 mod options;
-mod utils;
+use icu_locid::{
+    extensions::unicode::{key, Value},
+    Locale,
+};
+use num_bigint::BigInt;
+use num_traits::Num;
 pub(crate) use options::*;
-pub(crate) use utils::*;
+
+use crate::{
+    builtins::{
+        builder::BuiltInBuilder, options::get_option, string::is_trimmable_whitespace,
+        BuiltInConstructor, BuiltInObject, IntrinsicObject,
+    },
+    context::intrinsics::{Intrinsics, StandardConstructor, StandardConstructors},
+    js_string,
+    object::{
+        internal_methods::get_prototype_from_constructor, FunctionObjectBuilder, JsFunction,
+        ObjectInitializer,
+    },
+    property::{Attribute, PropertyDescriptor},
+    realm::Realm,
+    string::common::StaticJsStrings,
+    value::PreferredType,
+    Context, JsArgs, JsData, JsNativeError, JsObject, JsResult, JsString, JsSymbol, JsValue,
+    NativeFunction,
+};
+
+use super::{
+    locale::{canonicalize_locale_list, resolve_locale, supported_locales, validate_extension},
+    options::{coerce_options_to_object, IntlOptions},
+    Service,
+};
+
+#[derive(Debug, Trace, Finalize, JsData)]
+// Safety: `NumberFormat` only contains non-traceable types.
+#[boa_gc(unsafe_empty_trace)]
+pub(crate) struct NumberFormat {
+    locale: Locale,
+    formatter: FixedDecimalFormatter,
+    numbering_system: Option<Value>,
+    unit_options: UnitFormatOptions,
+    digit_options: DigitFormatOptions,
+    notation: Notation,
+    use_grouping: GroupingStrategy,
+    sign_display: SignDisplay,
+    bound_format: Option<JsFunction>,
+}
+
+impl NumberFormat {
+    /// [`FormatNumeric ( numberFormat, x )`][full] and [`FormatNumericToParts ( numberFormat, x )`][parts].
+    ///
+    /// The returned struct implements `Writable`, allowing to either write the number as a full
+    /// string or by parts.
+    ///
+    /// [full]: https://tc39.es/ecma402/#sec-formatnumber
+    /// [parts]: https://tc39.es/ecma402/#sec-formatnumbertoparts
+    fn format<'a>(&'a self, value: &'a mut FixedDecimal) -> FormattedFixedDecimal<'a> {
+        // TODO: Missing support from ICU4X for Percent/Currency/Unit formatting.
+        // TODO: Missing support from ICU4X for Scientific/Engineering/Compact notation.
+
+        self.digit_options.format_fixed_decimal(value);
+        value.apply_sign_display(self.sign_display);
+
+        self.formatter.format(value)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(in crate::builtins::intl) struct NumberFormatLocaleOptions {
```

Comment:
question: couldn't this be `pub(super)` vs `pub(in crate::builtins::intl)`? Is there a benefit to using the latter over the former?

Reply:
True.